### PR TITLE
Feature/231 - New experiment component

### PR DESCRIPTION
**Pull request by vidhya-metacell**
1.  New experiment creation
2. Required validation for experiment name before submission
3. Validation on values field for list type
4. Range fields set to type number
5. Create  new experiment dialog
6. List/range toggle
7. Group add/remove
8. Add Parameter
9. Remove Parameter
10. Autocomplete to first level of keys

### CODE REVIEW

**Comment by lrebscher** [comment]
Scrolling begins although there is still some space available
Bildschirmfoto 2021-04-06 um 11 01 09

**Comment by lrebscher** [comment]
Same as with the *Create New Experiment* button, on larger screens the  *Add Parameter* seems to me to be too far away from the actual Parameters list, would expect it to be directly below the parameters list

Bildschirmfoto 2021-04-06 um 11 00 44

**Comment by lrebscher** [comment]
Parameter selector field has different hint colors and selection color compared to other fields, field height and positioning of value looks off too in comparison to Type field. 
-> Keep in mind for #235 

Bildschirmfoto 2021-04-06 um 11 04 08
Bildschirmfoto 2021-04-06 um 11 05 50

**Comment by lrebscher** [comment]
Scrolling begins although there is some space left Bildschirmfoto 2021-04-06 um 11 01 09

File: webapp/components/experiments/EditExperiment.js [hunk 1]
```diff
@@ -1,43 +1,221 @@
...
+    },
+    '& .editExperimentFooter': {
+      position: 'absolute',
+      left: 0,
+      bottom: 0,
+      right: 0,
+      display: 'flex',
+      justifyContent: 'space-between',
+      alignItems: 'center',
+      background: '#434343',
```
Comment (lrebscher) [inline webapp/components/experiments/EditExperiment.js:147]: that color is defined in `variables.less`& referenced in `@background_color_body_0` in `colors.less`
Make sure to reuse these definitions & please check for any hardcoded color in here

File: webapp/components/experiments/EditExperiment.js [hunk 2]
```diff
@@ -49,21 +227,242 @@ const EditExperiment = () => {
...
+
+  const handleParamSelection = (val, parameter, index) => {
+    const newParam = parameter.inGroup ?  [...groupParameters] : [...parameters];
+    newParam[index] = { ...parameter, 'mapsTo': val };
+    parameter.inGroup ? setGroupParameters(newParam) : setParameters(newParam);
+  }
+
+  const ParameterMenu = (props) => {
+    const { parameter, index } = props;
+    const [anchorEl, setAnchorEl] = React.useState(null);
```
Comment (lrebscher) [inline webapp/components/experiments/EditExperiment.js:257]: `useState` is imported, don't have to use `React.` notation

**Comment by paulfaivret** [comment]
> Same as with the _Create New Experiment_ button, on larger screens the _Add Parameter_ seems to me to be too far away from the actual Parameters list, would expect it as a user to be directly below the parameters list
> 
> @paulfaivret would like to get your opinion on it
> 
> Bildschirmfoto 2021-04-06 um 11 00 44

Fixed it : the link is now in the main container, below the existing parameters.

File: webapp/components/experiments/Experiments.js [hunk 3]
```diff
@@ -126,9 +133,8 @@ const Experiments = (props) => {
   } = props;
 
   const POLL_INTERVAL = 1000;
-
   useEffect(getExperiments, []);
-  useInterval(getExperiments, POLL_INTERVAL);
+  // useInterval(getExperiments, POLL_INTERVAL);
```
Comment (lrebscher) [inline webapp/components/experiments/Experiments.js:137]: Revert comment

File: webapp/components/experiments/EditExperiment.js [hunk 4]
```diff
@@ -49,21 +227,242 @@ const EditExperiment = () => {
...
+          classes={{ paper: 'MuiPopover-experiment' }}
+          open={Boolean(anchorEl)}
+          onClose={handleClose}
+          anchorOrigin={{ vertical: 'bottom', horizontal: 'right' }}
+          transformOrigin={{ vertical: 'top', horizontal: 'right' }}
+        >
+          { parameter.inGroup ?
+            <MenuItem onClick={() => removeFromGroup(index)}>Remove from group</MenuItem> : <MenuItem onClick={() => addToGroup(index)}>Add to Group</MenuItem>
+          }
+          <MenuItem onClick={() => removeParamter(index, parameter)}>Delete</MenuItem>
```
Comment (lrebscher) [inline webapp/components/experiments/EditExperiment.js:285]: `removeParamter` -> `removeParameter`

File: webapp/components/experiments/EditExperiment.js [hunk 5]
```diff
@@ -49,21 +227,242 @@ const EditExperiment = () => {
...
+    parameter.inGroup ? setGroupParameters(newParam) : setParameters(newParam);
+  };
+
+  const handleParamSelection = (val, parameter, index) => {
+    const newParam = parameter.inGroup ?  [...groupParameters] : [...parameters];
+    newParam[index] = { ...parameter, 'mapsTo': val };
+    parameter.inGroup ? setGroupParameters(newParam) : setParameters(newParam);
+  }
+
+  const ParameterMenu = (props) => {
```
Comment (lrebscher) [inline webapp/components/experiments/EditExperiment.js:255]: Would avoid to define nested components, makes it less readable & hard to reuse, the component is tightly coupled & already quite large. 

One example for tight coupling: you define `handleClose` in `ParameterMenu` but use it e.g. in `addToGroup` method that is define outside of this scope, my IDE marks this, it only works atm due to the order of the function declarations. And the other way around, the inner component calls method of the outer component

Avoid large render methods that have more than ~100 lines of code & split this up into functions to increase readability & abstraction.

You can move components that have their own state to a separate file and smaller stateless components to the global level in the same file

File: webapp/components/experiments/EditExperiment.js [hunk 6]
```diff
@@ -49,21 +227,242 @@ const EditExperiment = () => {
...
+        </Box>
+        <Box className="editExperimentContent">
+          <Box mb={2} className="editExperimentDefault">
+            <Box mb={2} className="editExperimentBreadcrumb">
+              <Typography variant="body2">Parameters</Typography>
+            </Box>
+            { groupParameters.length > 0 && (
+              <Box mb={2} className="editExperimentGroup">
+                <Box mb={2} className="editExperimentBreadcrumb">
+                  <Typography variant="body2">Groupped Parameters</Typography>
```
Comment (lrebscher) [inline webapp/components/experiments/EditExperiment.js:421]: Groupped -> Grouped

File: webapp/components/experiments/EditExperiment.js [hunk 7]
```diff
@@ -49,21 +227,242 @@ const EditExperiment = () => {
...
+                </Box>
+                <Box className="editExperimentRow scrollbar scrollchild">
+                  {groupParameters.map((parameter, index) => (
+                    parameterRow(parameter, index)
+                  ))}
+                </Box>
+                { groupParameters.length === 1 && <Box className="editExpermentWarning">
```
Comment (lrebscher) [inline webapp/components/experiments/EditExperiment.js:428]: editExpermentWarning -> editExperimentWarning

File: webapp/components/experiments/EditExperiment.js [hunk 8]
```diff
@@ -49,21 +227,242 @@ const EditExperiment = () => {
...
+                  <Typography variant="caption">Warning: You need at least two parameters for the grouping to work.</Typography>
```
Comment (lrebscher) [inline webapp/components/experiments/EditExperiment.js:429]: Would extract any dialog content & hardcoded string to `constants.js`

File: webapp/components/experiments/ExperimentDialog.js [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+  return (
+    <Dialog
+      fullWidth
+      maxWidth="sm"
+      open={open}
+      onClose={() => dialogClose(false)}
+    >
+      <DialogTitle>Create New Experiment</DialogTitle>
+      <DialogContent>
+        <Typography variant="subtitle" className={classes.heading}>The new experiment will replace the current experiment in design, are you sure you want to proceed?</Typography>
```
Comment (lrebscher) [inline webapp/components/experiments/ExperimentDialog.js:34]: Extract to `constant.js`

File: webapp/components/experiments/ExperimentDialog.js [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+  } = props;
+
+  return (
+    <Dialog
+      fullWidth
+      maxWidth="sm"
+      open={open}
+      onClose={() => dialogClose(false)}
+    >
+      <DialogTitle>Create New Experiment</DialogTitle>
```
Comment (lrebscher) [inline webapp/components/experiments/ExperimentDialog.js:32]: Extract to constant.js

File: webapp/components/experiments/ExperimentDialog.js [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+      <DialogContent>
+        <Typography variant="subtitle" className={classes.heading}>The new experiment will replace the current experiment in design, are you sure you want to proceed?</Typography>
+      </DialogContent>
+      <DialogActions>
+        <Button
+          onClick={() => dialogClose(false)}
+          className={classes.cancel}
+          key="CANCEL"
+        >
+          CANCEL
```
Comment (lrebscher) [inline webapp/components/experiments/ExperimentDialog.js:42]: This is a good example why hardcoded strings lead to inconsistency, we already use "Cancel" in lowercase for other dialogs, now we introduce "CANCEL".

We should have one place where we define what the label for cancel button is and reuse that across all dialogs

File: webapp/components/experiments/ExperimentDialog.js [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+import { withStyles } from '@material-ui/core/styles';
+
+const useStyles = () => ({
+  root: {
+    '& .cancel': { marginRight: 10 },
+    '&. heading': { color: 'white' },
+  },
+});
+
+const ExperimentDialog = (props) => {
```
Comment (lrebscher) [inline webapp/components/experiments/ExperimentDialog.js:17]: Please generalize `DeleteDialogBox.js`, it has the same structure

File: webapp/components/experiments/Experiments.js [hunk 13]
```diff
@@ -154,80 +160,97 @@ const Experiments = (props) => {
...
-            <Table aria-label="simple table">
-              <TableBody>
-                {experiments.map((experiment) => (
-                  <TableRow key={experiment?.name}>
-                    <TableCell component="th" scope="row">
-                      <Button>
-                        <Typography variant="h6" className="experimentHead">
-                          {experiment?.name}
+    <>
+    { list ?
```
Comment (lrebscher) [inline webapp/components/experiments/Experiments.js:174]: Ternary operators increase readability for short conditional checks `active ? doThat() : doThis()`, ideally it should fit into one line, but avoid using it across 100 lines, there is too much detail in this check

rather do: `list ? <Experiments/> : <EditExperiment/>`

Also keep in mind that we have more than two options, the detail view of an Experiment is another component

File: webapp/components/experiments/Experiments.js [hunk 14]
```diff
@@ -154,80 +160,97 @@ const Experiments = (props) => {
...
+      <GridLayout className={classes.root}>
```
Comment (lrebscher) [inline webapp/components/experiments/Experiments.js:175]: I would suggest to extract the Experiment list to a separate component with its own state, the component that decides which view to show (list, detail, edit, ..) should not know of the experiment list state and its actions, ensures also a valid state for the Experiment list by design

File: webapp/css/netpyne.less [hunk 15]
```diff
@@ -593,4 +593,78 @@ div.flexlayout__layout {
 @keyframes spin {
   0% { transform: rotate(0deg); }
   100% { transform: rotate(360deg); }
-}
\ No newline at end of file
+}
+.editExperimentContainer {
```
Comment (lrebscher) [inline webapp/css/netpyne.less:597]: Why use `withStyles` and `netpyne.less` at the same time to style `EditExperiment.js`?

**Comment by lrebscher** [comment]
We have inconsistency in the input type for decimal/float type.

The range field doesn't allow dot notation for decimal (e.g. 1.2), you have to use comma (1,2)
the Values field separates values by ",", in order to use decimal values you have to use dot.

The existing fields in Netpyne use dot notation, one can see this by importing Tutorial 1 & opening the Connectivity Rules tab. "1,2" is interpreted as a tuple (1,2).

To summarize, we only support dot notation for decimals, the range fields must support this as well

File: webapp/components/experiments/Experiments.js [hunk 16]
```diff
@@ -21,14 +21,21 @@ import {
...
+import DeleteDialogBox from '../general/DeleteDialogBox';
+import { experimentGrey, experimentInputColor } from '../../theme';
 
 const useStyles = theme => ({
   root: {
+    '& .scrollDiv': {
+      overflow: 'auto',
+    },
     '& .MuiTableContainer-root': {
+      maxHeight: 'calc(100% -60vh)',
```
Comment (filippomc) [inline webapp/components/experiments/Experiments.js:37]: this doesn't smell right. I'd expect either 40vh or percent on the container

File: webapp/css/variables.less [hunk 17]
```diff
@@ -14,6 +14,7 @@
 @colorLink: #b9c680;
 
 @radius: 5px;
+@borderRadius: 4px;
```
Comment (filippomc) [inline webapp/css/variables.less:17]: why 2 radiuses? Should be the same everywhere

**Review by filippomc — COMMENTED** [review]
Looks good to me, added two minor comments

File: webapp/theme.js [hunk 18]
```diff
@@ -26,8 +26,16 @@ export const {
   bgInputs,
   gutter,
   radius,
+  borderRadius,
   canvasBgDark,
   canvasBgLight,
+  experimentAutomcompleteBorder,
```
Comment (lrebscher) [inline webapp/theme.js:32]: experimentAutocompleteBorder
experimentLabelColor
experimentSvgColor

**Review by lrebscher — APPROVED** [review]
👍
